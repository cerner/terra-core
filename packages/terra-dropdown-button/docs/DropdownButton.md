# Terra Dropdown Button

The Terra Dropdown Button is a button that, when pressed, opens a dropdown with more buttons.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-dropdown-button`

## Component Features

 * [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
 * [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
 * [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
 * [LTR/RTL Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#ltr--rtl-support)
