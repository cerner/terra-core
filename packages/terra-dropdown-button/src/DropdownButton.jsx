import React from 'react';
import PropTypes from 'prop-types';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import * as KeyCode from 'keycode-js';
import DropdownButtonBase from './_DropdownButtonBase';
import styles from './DropdownButton.module.scss';
import Item from './Item';
import SplitButton, { Variants as SplitButtonVariants } from './SplitButton';

const cx = classNamesBind.bind(styles);

const Variants = {
  NEUTRAL: 'neutral',
  EMPHASIS: 'emphasis',
  GHOST: 'ghost',
};

const propTypes = {
  /**
   * The options to display in the dropdown. Should be comprised of the subcomponent `Item`.
   */
  children: PropTypes.node.isRequired,
  /**
   * Determines whether the button should have block styling applied. The dropdown list will match the component's width.
   */
  isBlock: PropTypes.bool,
  /**
   * Whether or not the button has reduced padding for use in tables and single-row lists.
   */
  isCompact: PropTypes.bool,
  /**
   * Determines whether the button should be disabled.
   */
  isDisabled: PropTypes.bool,
  /**
   * Sets the text that will be shown on the dropdown button.
   */
  label: PropTypes.string.isRequired,
  /**
   * Sets the styles of the component, one of `neutral`, `emphasis`, or `ghost`.
   */
  variant: PropTypes.oneOf(['neutral', 'emphasis', 'ghost']),
};

const defaultProps = {
  isBlock: false,
  isCompact: false,
  isDisabled: false,
  variant: 'neutral',
};

class DropdownButton extends React.Component {
  constructor(props) {
    super(props);

    this.handleDropdownButtonClick = this.handleDropdownButtonClick.bind(this);
    this.handleDropdownRequestClose = this.handleDropdownRequestClose.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.setButtonNode = this.setButtonNode.bind(this);
    this.getButtonNode = this.getButtonNode.bind(this);
    this.setListNode = this.setListNode.bind(this);
    this.toggleDropDown = this.toggleDropDown.bind(this);
    this.state = { isOpen: false, isActive: false, openedViaKeyboard: false };
  }

  setListNode(element) {
    this.dropdownList = element;
  }

  setButtonNode(node) {
    this.buttonNode = node;
  }

  getButtonNode() {
    return this.buttonNode;
  }

  toggleDropDown(event) {
    this.setState(prevState => ({ isOpen: !prevState.isOpen }));
    // See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Button#Clicking_and_focus
    // Button on Firefox, Safari and IE running on OS X does not receive focus when clicked.
    // This will put focus on the button when clicked.
    event.currentTarget.focus();
  }

  handleDropdownButtonClick(event) {
    if (this.state.isOpen) {
      this.setState({ openedViaKeyboard: false });
    }
    this.toggleDropDown(event);
  }

  handleDropdownRequestClose(callback) {
    const onSelectCallback = typeof callback === 'function' ? callback : undefined;
    this.setState({ isOpen: false, openedViaKeyboard: false, isActive: false }, onSelectCallback);
  }

  handleKeyDown(event) {
    if (this.state.isOpen && event.keyCode === KeyCode.KEY_ESCAPE) {
      event.stopPropagation();
    }
    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
      // In FireFox active styles don't get applied on space
      this.setState({ isActive: true, openedViaKeyboard: true });
      /*
        Prevent the callback from being called repeatedly if the RETURN or SPACE key is held down.
        The keyDown event of native html button triggers Onclick() event on RETURN or SPACE key press.
        where holding RETURN key for longer time will call dropdownClick() event repeatedly which would cause
        the dropdown to open and close itself.
      */
      event.preventDefault();
    } else if (event.keyCode === KeyCode.KEY_DOWN && this.state.isOpen && !this.state.openedViaKeyboard) {
      // set focus to first list element on down arrow key press only when dropdown is opened by mouse click.
      const listOptions = this.dropdownList.querySelectorAll('[data-terra-dropdown-list-item]');
      listOptions[0].focus();
      // prevent handleFocus() callback of DropdownList.
      event.preventDefault();
    } else if (event.keyCode === KeyCode.KEY_UP && this.state.isOpen && !this.state.openedViaKeyboard) {
      // set focus to last list element on up arrow key press only when dropdown is opened by mouse click
      const listOptions = this.dropdownList.querySelectorAll('[data-terra-dropdown-list-item]');
      listOptions[listOptions.length - 1].focus();
      event.preventDefault();
    } else if (event.keyCode === KeyCode.KEY_TAB) {
      this.handleDropdownRequestClose();
    }
  }

  handleKeyUp(event) {
    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
      event.preventDefault();
      this.setState({ isActive: false });
      this.toggleDropDown(event);
    }
  }

  render() {
    const {
      children,
      isBlock,
      isCompact,
      isDisabled,
      label,
      variant,
      ...customProps
    } = this.props;

    const theme = this.context;

    const { isOpen, isActive, openedViaKeyboard } = this.state;

    const classnames = cx(
      'dropdown-button',
      variant,
      { 'is-active': isOpen || isActive },
      { 'is-block': isBlock },
      { 'is-compact': isCompact },
      /* This needs to match terra-hookshot's react-onclickoutside ignore classname or clicking the caret with
        the dropdown open will cause the dropdown to close and reopen
      */
      { 'ignore-react-onclickoutside': isOpen },
      theme.className,
    );

    return (
      <DropdownButtonBase
        {...customProps}
        items={children}
        isOpen={isOpen}
        isBlock={isBlock}
        isCompact={isCompact}
        isDisabled={isDisabled}
        requestClose={this.handleDropdownRequestClose}
        openedViaKeyboard={openedViaKeyboard}
        buttonRef={this.getButtonNode}
        refCallback={this.setListNode}
      >
        <button
          type="button"
          className={classnames}
          onClick={this.handleDropdownButtonClick}
          onKeyDown={this.handleKeyDown}
          onKeyUp={this.handleKeyUp}
          disabled={isDisabled}
          tabIndex={isDisabled ? '-1' : undefined}
          aria-disabled={isDisabled}
          aria-expanded={isOpen}
          aria-haspopup="menu"
          ref={this.setButtonNode}
        >
          <span className={cx('dropdown-button-text')}>{label}</span>
          <span className={cx('caret-icon')} />
        </button>
      </DropdownButtonBase>
    );
  }
}

DropdownButton.propTypes = propTypes;
DropdownButton.defaultProps = defaultProps;
DropdownButton.contextType = ThemeContext;

export default DropdownButton;
export {
  Item, Variants, SplitButton, SplitButtonVariants,
};
