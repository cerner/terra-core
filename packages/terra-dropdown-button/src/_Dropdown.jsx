import React from 'react';
import PropTypes from 'prop-types';
import FocusTrap from 'focus-trap-react';
import Hookshot from 'terra-hookshot';
import DropdownList from './_DropdownList';

const propTypes = {
  /**
   * Callback to tell the parent it should close the dropdown
   */
  requestClose: PropTypes.func.isRequired,
  /**
   * Ref to attach the dropdown to
   */
  targetRef: PropTypes.func.isRequired,
  /**
   * Whether or not the dropdown should be open
   */
  isOpen: PropTypes.bool.isRequired,
  /**
   * The Options that should be displayed in the dropdown
   */
  children: PropTypes.node.isRequired,
  /**
   * @private
   * Width to set the dropdown to. Used when `isBlock` is true. Unset means to autosize.
   */
  width: PropTypes.string,
};

const Dropdown = ({
  requestClose, isOpen, targetRef, children, width,
}) => (
  <Hookshot
    isOpen={isOpen}
    isEnabled
    targetRef={targetRef}
    attachmentBehavior="flip"
    contentAttachment={{ vertical: 'top', horizontal: 'start' }}
    targetAttachment={{ vertical: 'bottom', horizontal: 'start' }}
  >
    <Hookshot.Content
      onEsc={requestClose}
      onOutsideClick={requestClose}
    >
      <FocusTrap focusTrapOptions={{ returnFocusOnDeactivate: true, clickOutsideDeactivates: true }}>
        <DropdownList
          requestClose={requestClose}
          width={width}
        >
          {children}
        </DropdownList>
      </FocusTrap>
    </Hookshot.Content>
  </Hookshot>
);

Dropdown.propTypes = propTypes;

export default Dropdown;
