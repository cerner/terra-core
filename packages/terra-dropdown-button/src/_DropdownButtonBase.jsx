import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import Dropdown from './_Dropdown';
import styles from './_DropdownButtonBase.module.scss';

const cx = classNamesBind.bind(styles);

const propTypes = {
  /**
   * The button or buttons that are the primary option and caret
   */
  children: PropTypes.node.isRequired,
  /**
   * The options to display in the dropdown. Should be of type `Button`.
   */
  items: PropTypes.node.isRequired,
  /**
   * Whether or not the dropdown is open
   */
  isOpen: PropTypes.bool.isRequired,
  /**
   * Callback to tell the parent it should close the dropdown
   */
  requestClose: PropTypes.func.isRequired,
  /**
   * Determines whether the component should have block styles applied. The dropdown will match the component's width.
   */
  isBlock: PropTypes.bool,
  /**
   * Whether or not the button has reduced padding for use in tables and single-row lists.
   */
  isCompact: PropTypes.bool,
  /**
   * Determines whether the primary button and expanding the dropdown should be disabled.
   */
  isDisabled: PropTypes.bool,
  /**
   * Whether or not dropdown is opened using keyboard.
   */
  openedViaKeyboard: PropTypes.bool,
  /**
   * Callback for reference of the dropdown button
   */
  buttonRef: PropTypes.func,
  /**
   * Ref callback for the dropdown list DOM element.
   */
  refCallback: PropTypes.func,
};

const defaultProps = {
  isBlock: false,
  isCompact: false,
  isDisabled: false,
  openedViaKeyboard: false,
};

class DropdownButtonBase extends React.Component {
  constructor(props) {
    super(props);

    this.setButtonWrapperRef = this.setButtonWrapperRef.bind(this);
    this.getButtonWrapperRef = this.getButtonWrapperRef.bind(this);

    this.buttonWrapperRef = null;
  }

  setButtonWrapperRef(ref) {
    this.buttonWrapperRef = ref;
  }

  getButtonWrapperRef() {
    return this.buttonWrapperRef;
  }

  render() {
    const {
      children,
      items,
      isOpen,
      requestClose,
      isBlock,
      isCompact,
      isDisabled,
      openedViaKeyboard,
      buttonRef,
      refCallback,
      ...customProps
    } = this.props;

    const theme = this.context;

    const DropdownButtonClassNames = classNames(
      cx(
        'dropdown-button-base',
        { 'is-block': isBlock },
        { 'is-compact': isCompact },
        { disabled: isDisabled },
        theme.className,
      ),
      customProps.className,
    );

    let calcWidth;
    if (this.buttonWrapperRef && isBlock) {
      calcWidth = `${this.buttonWrapperRef.getBoundingClientRect().width}px`;
    }

    return (
      <div
        {...customProps}
        className={DropdownButtonClassNames}
        ref={this.setButtonWrapperRef}
      >
        {children}
        <Dropdown
          targetRef={this.getButtonWrapperRef}
          isOpen={isOpen}
          requestClose={requestClose}
          width={calcWidth}
          openedViaKeyboard={openedViaKeyboard}
          buttonRef={buttonRef}
          refCallback={refCallback}
        >
          {items}
        </Dropdown>
      </div>
    );
  }
}

DropdownButtonBase.propTypes = propTypes;
DropdownButtonBase.defaultProps = defaultProps;
DropdownButtonBase.contextType = ThemeContext;

export default DropdownButtonBase;
