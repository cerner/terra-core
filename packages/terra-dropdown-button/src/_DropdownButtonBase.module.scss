// Themes
@import './clinical-lowlight-theme/DropdownButtonBase.module';
@import './orion-fusion-theme/DropdownButtonBase.module';

:local {
  .dropdown-button-base {
    border-radius: var(--terra-dropdown-button-border-radius, 3px);
    box-shadow: var(--terra-dropdown-button-box-shadow);
    display: inline-flex;
    flex-direction: row;
    flex-wrap: nowrap;
    font-family: inherit;
    font-size: var(--terra-dropdown-button-font-size, 1rem);
    line-height: var(--terra-dropdown-button-line-height, 1.42857);

    &:disabled,
    &.disabled {
      opacity: var(--terra-dropdown-button-disabled-opacity, 0.3);
      pointer-events: none;
    }

    &.is-block {
      display: flex;
    }
  }
}
