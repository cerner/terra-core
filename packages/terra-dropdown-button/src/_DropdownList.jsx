import React from 'react';
import PropTypes from 'prop-types';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import * as KeyCode from 'keycode-js';
import Util from './_DropdownListUtil';
import styles from './_DropdownList.module.scss';

const cx = classNamesBind.bind(styles);

const propTypes = {
  /**
   * The Options that should be displayed in the dropdown
   */
  children: PropTypes.node.isRequired,
  /**
   * Callback to tell the parent it should close the dropdown
   */
  requestClose: PropTypes.func.isRequired,
  /**
   * @private
   * Width to set the dropdown to. Used when `isBlock` is true. Unset means to autosize.
   */
  width: PropTypes.string,
  /**
   * Ref callback for the dropdown list DOM element.
   */
  refCallback: PropTypes.func,
};

class DropdownList extends React.Component {
  constructor(props) {
    super(props);

    this.cloneChildren = this.cloneChildren.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.clearSearch = this.clearSearch.bind(this);
    this.changeFocusState = this.changeFocusState.bind(this);

    this.state = { focused: 0, active: -1 };

    this.searchString = '';
    this.pressed = false;
    this.listRef = null;
  }

  changeFocusState(index) {
    // the div inside the li is what is actually focusable so need to go 2 layers down
    this.listRef.childNodes[index].childNodes[0].focus();

    this.setState({ focused: index });
  }

  handleKeyDown(event) {
    const { keyCode } = event;
    const { focused } = this.state;
    if (keyCode === KeyCode.KEY_RETURN || keyCode === KeyCode.KEY_SPACE) {
      /*
        Prevent the callback from being called repeatedly if key is held down.
        The close dropdown request had to be moved to handleKeyUp to fix a firefox bug
        where choosing an item with spacebar if the dropdown caret was focused when opening the dropdown
        would cause the dropdown to reopen itself.
      */
      if (!this.pressed) {
        this.pressed = true;
        this.setState({ active: focused });
      }
      event.preventDefault();
    } else if (keyCode === KeyCode.KEY_DOWN) {
      if (!this.pressed) {
        this.changeFocusState(Util.findNext(this, this.state.focused));
      }
      event.preventDefault();
    } else if (keyCode === KeyCode.KEY_UP) {
      if (!this.pressed) {
        this.changeFocusState(Util.findPrevious(this, this.state.focused));
      }
      event.preventDefault();
    } else if (keyCode === KeyCode.KEY_HOME) {
      if (!this.pressed) {
        this.changeFocusState(0);
      }
      event.preventDefault();
    } else if (keyCode === KeyCode.KEY_END) {
      if (!this.pressed) {
        this.changeFocusState(Util.getChildArray(this).length - 1);
      }
      event.preventDefault();
    } else if (keyCode === KeyCode.KEY_TAB) {
      this.props.requestClose();
      event.preventDefault();
    } else if (keyCode >= 48 && keyCode <= 90) {
      this.searchString = this.searchString.concat(String.fromCharCode(keyCode));
      clearTimeout(this.searchTimeout);
      this.searchTimeout = setTimeout(this.clearSearch, 500);
      let newFocused = Util.findWithStartString(this, this.searchString);

      if (newFocused === -1) {
        newFocused = this.state.focused;
      }
      this.changeFocusState(newFocused);
    }
    event.stopPropagation();
  }

  handleKeyUp(event) {
    const { keyCode } = event;
    if (keyCode === KeyCode.KEY_RETURN || keyCode === KeyCode.KEY_SPACE) {
      this.setState({ active: undefined });
      event.preventDefault();
      if (this.pressed) {
        const item = Util.findByIndex(this, this.state.focused);
        this.props.requestClose(() => {
          item.props.onSelect(event, item.props.metaData);
        });
      }
      this.pressed = false;
    }
  }

  /**
   * Keeps internal state synced with what is focused on the page incase
   * something such as a screen reader moves focus
   * @param {React.FocusEvent} event the focus event
   */
  handleFocus(event) {
    const index = Util.findIndexByValue(this, event.target.innerText);

    if (index !== -1) {
      this.setState({ focused: index });
    }
  }

  /**
   * Clears the keyboard search.
   */
  clearSearch() {
    this.searchString = '';
    clearTimeout(this.searchTimeout);
    this.searchTimeout = null;
  }

  /**
   * Tells children whether or not they are active and gives them a function to close the dropdown
   * @return {Array<React.ReactNode>} the array of children
   */
  cloneChildren() {
    return React.Children.map(this.props.children, (child, index) => React.cloneElement(child, {
      isActive: index === this.state.active,
      requestClose: this.props.requestClose,
      'data-terra-dropdown-list-item': true,
    }));
  }

  /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
  render() {
    const {
      width,
      refCallback,
    } = this.props;

    const theme = this.context;

    return (
      <ul
        className={cx('dropdown-list', theme.className)}
        // eslint-disable-next-line react/forbid-dom-props
        style={{ width }}
        ref={(ref) => {
          if (refCallback) {
            refCallback(ref);
          }
          this.listRef = ref;
        }}
        onFocus={this.handleFocus}
        onKeyDown={this.handleKeyDown}
        onKeyUp={this.handleKeyUp}
        role="menu"
      >
        {this.cloneChildren()}
      </ul>
    );
  }
}

DropdownList.propTypes = propTypes;
DropdownList.contextType = ThemeContext;

export default DropdownList;
