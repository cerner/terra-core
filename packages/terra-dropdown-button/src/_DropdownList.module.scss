// Themes
@import './clinical-lowlight-theme/DropdownList.module';
@import './orion-fusion-theme/DropdownList.module';

:local {
  .dropdown-list {
    background-color: var(--terra-dropdown-button-list-background-color, #fff);
    border-radius: var(--terra-dropdown-button-list-border-radius, 3px);
    box-shadow: var(--terra-dropdown-button-list-box-shadow, 0 4px 8px 0 rgba(0, 0, 0, 0.35), 0 4px 3px -3px rgba(0, 0, 0, 0.5));
    list-style: none;
    margin: 0;
    min-width: var(--terra-dropdown-button-list-min-width, 10rem);
    outline: none;
    overflow: hidden;
    padding-bottom: var(--terra-dropdown-button-list-padding-bottom, 0.142857rem);
    padding-left: 0;
    padding-right: 0;
    padding-top: var(--terra-dropdown-button-list-padding-top, 0.42857rem);
  }
}
