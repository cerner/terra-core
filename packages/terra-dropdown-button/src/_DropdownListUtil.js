import React from 'react';

class DropdownListUtil {
  /**
   * Compares the value of two strings to determine case insensitive equality.
   * @param {string|undefined} a - The initial value to compare.
   * @param {string|undefined} b - The secondary value to compare.
   * @return {boolean} - True if the values are equal.
   */
  static isEqual(a, b) {
    if (a === undefined || a === null || b === undefined || b === null) {
      return false;
    }
    return a.toString().toLowerCase() === b.toString().toLowerCase();
  }

  /**
   * Gets the children of a node as an array
   * @param {React.ReactNode} object Node to get children of
   * @return {Array<React.ReactNode>} the array of children
   */
  static getChildArray(object) {
    return React.Children.toArray(object.props.children);
  }

  /**
   * Finds the option matching the requested label.
   * @param {React.ReactNode} object - The dropdown containing the options.
   * @param {string} label - The label of the target option.
   * @return {React.ReactNode|undefined} - The option. Returns undefined if not found.
   */
  static findByValue(object, label) {
    return DropdownListUtil.getChildArray(object).find(({ props }) => DropdownListUtil.isEqual(props.label, label));
  }

  /**
   * Finds the index of the option matching the requested label.
   * @param {React.ReactNode} object - The dropdown containing the options.
   * @param {string} label - The label of the target option.
   * @return {number} - The index of the option
   */
  static findIndexByValue(object, label) {
    return DropdownListUtil.getChildArray(object).findIndex(({ props }) => DropdownListUtil.isEqual(props.label, label));
  }

  static findByIndex(object, index) {
    return DropdownListUtil.getChildArray(object)[index];
  }

  /**
   * Finds the first option in the dropdown that starts with the string.
   * @param {React.ReactNode} object - The dropdown.
   * @param {string} string - The start string.
   * @return {number} - The index of the first option that starts with the provided string.
   */
  static findWithStartString(object, string) {
    return DropdownListUtil.getChildArray(object).findIndex(opt => (
      opt.props.label || '').toLowerCase().startsWith(string.toLowerCase()));
  }

  /**
   * Finds the index of the option following the active option.
   * @param {React.ReactNode} object - The dropdown containing the options.
   * @param {number} index - The index of the active option.
   * @return {number} - The index of the next option.
   */
  static findNext(object, index) {
    const options = DropdownListUtil.getChildArray(object);
    const next = Math.min(index + 1, options.length - 1);
    return Math.max(0, next);
  }

  /**
   * Finds the index of the option preceding the active option.
   * @param {React.ReactNode} object - The dropdown containing the options.
   * @param {number} index - The index of the active option.
   * @return {number} - The index of the previous option.
   */
  static findPrevious(object, index) {
    const options = DropdownListUtil.getChildArray(object);
    const previous = Math.max(index - 1, 0);
    return Math.min(previous, options.length - 1);
  }
}

export default DropdownListUtil;
