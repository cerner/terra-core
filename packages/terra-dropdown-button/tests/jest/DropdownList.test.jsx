import React from 'react';
import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';

import DropdownList from '../../src/_DropdownList';
import { Item } from '../../src/DropdownButton';

describe('Dropdown List', () => {
  it('renders a default dropdown list', () => {
    const wrapper = shallow(
      <DropdownList requestClose={() => {}}>
        <Item label="1st Option" onSelect={() => {}} />
      </DropdownList>,
    );

    expect(wrapper).toMatchSnapshot();
    // ensures both data attributes are added when list contains single item
    const item = wrapper.find('Item');
    expect(item.props()).toHaveProperty('data-terra-dropdown-list-item');
  });

  it('renders a dropdown list with a set width', () => {
    const wrapper = shallow(
      <DropdownList requestClose={() => {}} width="440px">
        <Item label="1st Option" onSelect={() => {}} />
      </DropdownList>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a dropdown list with multiple children', () => {
    const wrapper = shallow(
      <DropdownList requestClose={() => {}}>
        <Item label="1st Option" onSelect={() => {}} />
        <Item label="2nd Option" onSelect={() => {}} />
        <Item label="3rd Option" onSelect={() => {}} />
      </DropdownList>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a dropdown list a non-default focused option', () => {
    const wrapper = shallow(
      <DropdownList requestClose={() => {}}>
        <Item label="1st Option" onSelect={() => {}} />
        <Item label="2nd Option" onSelect={() => {}} />
        <Item label="3rd Option" onSelect={() => {}} />
      </DropdownList>,
    );
    wrapper.setState({ focused: '2nd Option' });

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a dropdown list an active option', () => {
    const wrapper = shallow(
      <DropdownList requestClose={() => {}}>
        <Item label="1st Option" onSelect={() => {}} />
        <Item label="2nd Option" onSelect={() => {}} />
        <Item label="3rd Option" onSelect={() => {}} />
      </DropdownList>,
    );
    wrapper.setState({ active: '3rd Option' });

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a dropdown list an active and focused option', () => {
    const wrapper = shallow(
      <DropdownList requestClose={() => {}}>
        <Item label="1st Option" onSelect={() => {}} />
        <Item label="2nd Option" onSelect={() => {}} />
        <Item label="3rd Option" onSelect={() => {}} />
      </DropdownList>,
    );
    wrapper.setState({ focused: '2nd Option', active: '2nd Option' });

    expect(wrapper).toMatchSnapshot();
  });

  it('correctly applies the theme context className', () => {
    const wrapper = mount(
      <ThemeContextProvider theme={{ className: 'orion-fusion-theme' }}>
        <DropdownList requestClose={() => {}}>
          <Item label="1st Option" onSelect={() => {}} />
          <Item label="2nd Option" onSelect={() => {}} />
          <Item label="3rd Option" onSelect={() => {}} />
        </DropdownList>
      </ThemeContextProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
