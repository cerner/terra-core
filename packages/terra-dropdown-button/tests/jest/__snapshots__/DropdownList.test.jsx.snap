// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown List correctly applies the theme context className 1`] = `
<ThemeContextProvider
  theme={
    Object {
      "className": "orion-fusion-theme",
    }
  }
>
  <DropdownList
    requestClose={[Function]}
  >
    <ul
      className="dropdown-list orion-fusion-theme"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="menu"
      style={
        Object {
          "width": undefined,
        }
      }
    >
      <Item
        data-terra-dropdown-list-item={true}
        isActive={false}
        key=".0"
        label="1st Option"
        onSelect={[Function]}
        requestClose={[Function]}
      >
        <li
          role="presentation"
          tabIndex="-1"
        >
          <div
            className="item orion-fusion-theme"
            data-terra-dropdown-list-item={true}
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
          >
            1st Option
          </div>
        </li>
      </Item>
      <Item
        data-terra-dropdown-list-item={true}
        isActive={false}
        key=".1"
        label="2nd Option"
        onSelect={[Function]}
        requestClose={[Function]}
      >
        <li
          role="presentation"
          tabIndex="-1"
        >
          <div
            className="item orion-fusion-theme"
            data-terra-dropdown-list-item={true}
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
          >
            2nd Option
          </div>
        </li>
      </Item>
      <Item
        data-terra-dropdown-list-item={true}
        isActive={false}
        key=".2"
        label="3rd Option"
        onSelect={[Function]}
        requestClose={[Function]}
      >
        <li
          role="presentation"
          tabIndex="-1"
        >
          <div
            className="item orion-fusion-theme"
            data-terra-dropdown-list-item={true}
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
          >
            3rd Option
          </div>
        </li>
      </Item>
    </ul>
  </DropdownList>
</ThemeContextProvider>
`;

exports[`Dropdown List renders a default dropdown list 1`] = `
<ul
  className="dropdown-list"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="menu"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".0"
    label="1st Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
</ul>
`;

exports[`Dropdown List renders a dropdown list a non-default focused option 1`] = `
<ul
  className="dropdown-list"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="menu"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".0"
    label="1st Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".1"
    label="2nd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".2"
    label="3rd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
</ul>
`;

exports[`Dropdown List renders a dropdown list an active and focused option 1`] = `
<ul
  className="dropdown-list"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="menu"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".0"
    label="1st Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".1"
    label="2nd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".2"
    label="3rd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
</ul>
`;

exports[`Dropdown List renders a dropdown list an active option 1`] = `
<ul
  className="dropdown-list"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="menu"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".0"
    label="1st Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".1"
    label="2nd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".2"
    label="3rd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
</ul>
`;

exports[`Dropdown List renders a dropdown list with a set width 1`] = `
<ul
  className="dropdown-list"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="menu"
  style={
    Object {
      "width": "440px",
    }
  }
>
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".0"
    label="1st Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
</ul>
`;

exports[`Dropdown List renders a dropdown list with multiple children 1`] = `
<ul
  className="dropdown-list"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="menu"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".0"
    label="1st Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".1"
    label="2nd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
  <Item
    data-terra-dropdown-list-item={true}
    isActive={false}
    key=".2"
    label="3rd Option"
    onSelect={[Function]}
    requestClose={[Function]}
  />
</ul>
`;
